pretyping/evarconv.ml

```diff
@@ -133,7 +175,8 @@ let occur_rigidly (evk,_ as ev) evd t =
     | Proj (p, c) -> not (aux c)
     | Evar (evk',_) -> if Evar.equal evk evk' then raise Occur else false
     | Cast (p, _, _) -> aux p
-    | Lambda _ | LetIn _ -> false
+    | Lambda (na, t, b) -> aux b
+    | LetIn (na, _, _, b) -> aux b
```

Comment:
Can you explain this change please?
Why are you skipping the body of the let-in and the type of the lambda?

Reply:
It is an approximation: for lambdas and let-ins, short of doing reduction we don't know if an evar occurrence in the binding type or body of the let-in is indeed rigid if we reduced, so we only check the body. If it appears rigidly in the body then we know it would appear rigidly in the normal form as well.

Reply:
I'm sorry but I'm totally lost.

Why recurring under lambdas, but not in the branches of a match for example.
Also, if you don't go in the type of a lambda, how can this code "over approximate" occur check (that I guess looks everywhere).

Also if t is `(fix ....) t1... ev.. tn` it says true, while if t it is `def t1..ev..tn` is says false (`def` being a constant).
Both context are not necessarily rigid, so I really don't get what this code is supposed to do.

Finally, `Proj` flips the boolean: turning a regular projection into a primitive one changes what this code does?

It is good that you've add a comment, but I'm still puzzled and I don't understand how to explain what
this code does.

I looked at the bug you mention, it is "?X ?Y == X? (mk_pair (fst ?Y) (snd ?Y))". Can you explain the role played by this code? I expect it is `ev=?Y` and `t=(mk_pair (fst ?Y) (snd ?Y))`. When `fst` is primitive the `raise Occur` code path is reached, but if `fst` is a regular constant it is not.

Reply:
Sorry for the confusion, that function is tricky indeed. The comment was wrong. And the definition as well. What we want is an under-approximation of occur-check: it should return true only if an occur-check would return true on the normal form, but it might return false while the occur-check on the normal form would return true. So, the fix/cofix/meta cases were wrong: they are not definitely rigid heads. Proj is not either, unless it is not unfoldable, and same goes for constants. I adapted the code and the comment to reflect this.

In the example bug, occur_rigidly will now return false in both cases.

Reply:
I should be more obvious what the function does now.